dotnet-desktop-guide/net/winforms/controls-design/usercontrol-guidelines.md

```diff
@@ -0,0 +1,51 @@
+
+This article provides guidelines for creating user controls. We recommend that you follow these guidelines to make sure that you design a user control that's consistent with other Windows Forms controls.
+
+[!INCLUDE [desktop guide under construction](../../includes/desktop-guide-preview-note.md)]
+
+## Defining events
+
+Events commonly communicate state change and alert you to how the user is interacting with a Windows Forms control. For more information about events and delegates, see [Handle and raise events](/dotnet/standard/events/index).
+
+When defining your own events, follow these suggestions:
+
+- Use the <xref:System.EventHandler> event delegate when you define an event that doesn't have any associated data. Use the <xref:System.EventHandler%601> event delegate when you do have associated data.
+- Pass the control instance as the `sender` parameter.
+- Derive from <xref:System.EventArgs> and extend it with your data, when you raise an event with custom data.
+- Create a method named `On{EventName}` that raises the event, which is marked as `protected` and `virtual` (in C#) or `Protected` and `Overridable` (in Visual Basic).
+
+:::code language="csharp" source="./snippets/usercontrol-guidelines/csharp/UserControl1.cs" id="property_changed":::
+:::code language="vb" source="./snippets/usercontrol-guidelines/vb/UserControl1.vb" id="property_changed":::
+
+## Property changed events
+
+If you want your control to send notifications when a property changes, define an event named `{PropertyName}Changed`. This is the naming convention used in Windows Forms. The associated event delegate type for property-changed events is <xref:System.EventHandler>, and the event data type is <xref:System.EventArgs>. The base class <xref:System.Windows.Forms.Control> defines many property-changed events, such as <xref:System.Windows.Forms.Control.BackColorChanged>, <xref:System.Windows.Forms.Control.BackgroundImageChanged>, <xref:System.Windows.Forms.Control.FontChanged>, <xref:System.Windows.Forms.Control.LocationChanged>.  
+
+The same suggestions in the [Defining events](#defining-events) section apply here, as well.
+
+:::code language="csharp" source="./snippets/usercontrol-guidelines/csharp/UserControl1.cs" id="eventargs":::
+:::code language="vb" source="./snippets/usercontrol-guidelines/vb/UserControl1.vb" id="eventargs":::
+
+## Properties
+
+Control properties should support the Windows Forms Visual Designer. The **Properties** window interacts with control properties, and users expect to use this to change properties of the control. Either add the <xref:System.ComponentModel.DefaultValueAttribute> to the property, or create corresponding `Reset<PropertyName>` and `ShouldSerialize<PropertyName>` methods. For more information, see [DefaultValueAttribute](designer-properties-overview.md#defaultvalueattribute) and [Reset and ShouldSerialize](designer-properties-overview.md#reset-and-shouldserialize).
```

Comment:
```suggestion
Control properties should support the Windows Forms Visual Designer infrastructure. The **Properties** window interacts with control properties, and users expect to use this to change properties of the control. Either add the <xref:System.ComponentModel.DefaultValueAttribute> to the property, or create corresponding `Reset<PropertyName>` and `ShouldSerialize<PropertyName>` methods. For more information, see [DefaultValueAttribute](designer-properties-overview.md#defaultvalueattribute) and [Reset and ShouldSerialize](designer-properties-overview.md#reset-and-shouldserialize).
```

I am also wondering, if we should link a list of other attributes, which helps configuring the behavior in the Designer.

- **`[Browsable(bool)]`**: 
    - Specifies whether a property or event should be displayed in the Property Browser. A value of `false` hides the property or event.

- **`[DefaultValue(typeof(Type), "value")]`**:
    - Specifies the default value for a property. The WinForms designer uses this to determine whether the property should be serialized in the `InitializeComponent` method. If a property's current value matches its default, it typically isn't serialized.

- **`[Category("CategoryName")]`**:
    - Assigns the property to a specific category in the Property Browser. This helps in organizing properties.

- **`[Description("Description text")]`**:
    - Provides a brief description of the property, which is displayed at the bottom of the Property Browser when the property is selected.

- **`[DisplayName("Name")]`**:
    - Specifies a display name for a property or event, which can differ from the actual property or event name.

- **`[EditorBrowsable(EditorBrowsableState)]`**:
    - Determines if a property or method is visible in an editor, such as Visual Studio's IntelliSense. You can set it to `Never`, `Always`, or `Advanced`.

- **`[DesignerSerializationVisibility(DesignerSerializationVisibility)]`**:
    - Influences how or if a property is serialized at design-time. Possible values are `Visible` (the default), `Hidden` (not serialized), and `Content` (for properties that represent collections or objects).

- **`[Editor("EditorTypeName", "BaseEditorTypeName")]`**:
    - Specifies the editor to use for modifying the property's value. This can be used to provide a custom UI for property value editing.

- **`[TypeConverter(typeof(TypeConverter))]`**:
    - Specifies a type converter class for the property. Type converters can convert values between data types, provide values for a property, or indicate if a value supports a standard set of values.

- **`[RefreshProperties(RefreshProperties)]`**:
    - Indicates that changing this property requires refreshing other properties in the Property Browser. The values can be `All`, `Repaint`, or `None`.

- **`[FlagsAttribute()]`**:
    - Used with enumeration properties, allowing the combination of multiple enum values.

(Assembled by ChatGPT)